File: leetcode/easy/234_palindrome_linked_list.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val  = val
        self.next = next

# -----------------------------------------
# My Solution
#
# Time  Complexity: O(n)
# Space Complexity: O(n)
# -----------------------------------------
class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        vals = []

        node = head
        while node is not None:
            vals.append(node.val)
            node = node.next
        
        for i in range(len(vals) // 2):
            if vals[i] != vals[len(vals) - 1 - i]:
                return False
        
        return True

# -----------------------------------------
# Reverse List + Two Pointers
#
# Time  Complexity: O(n)
# Space Complexity: O(1)
# -----------------------------------------
# Ref: https://blog.csdn.net/coder_orz/article/details/51306985
class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        if head is None or head.next is None:
            return True

        # find the middle node (which would be slow after while loop)
        slow = fast = head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next

        slow = slow.next
        slow = self.reverse_list(slow)
        while slow:
            if head.val != slow.val:
                return False
            slow = slow.next
            head = head.next

        return True

    def reverse_list(self, head: ListNode) -> ListNode:
        prev_node = None
        curr_node = head
        while curr_node is not None:
            next_node = curr_node.next
            curr_node.next = prev_node

            prev_node = curr_node
            curr_node = next_node

        return prev_node

File: leetcode/easy/234_palindrome_linked_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// -----------------------------------------
// Model Solution: Reverse List + Two Pointers
//
// Time  Complexity: O(n)
// Space Complexity: O(1)
// -----------------------------------------
// n := number of nodes of linked list
// Ref: https://blog.csdn.net/coder_orz/article/details/51306985
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        ListNode* slow_pointer = head;
        ListNode* fast_pointer = head;
        while (fast_pointer && fast_pointer->next) {
            slow_pointer = slow_pointer->next;
            fast_pointer = fast_pointer->next->next;
        }

        ListNode* reversed_head = reverse_linked_list(slow_pointer);
        while (reversed_head) {
            if (head->val != reversed_head->val)
                return false;
            head = head->next;
            reversed_head = reversed_head->next;
        }
        return true;
    }

    ListNode* reverse_linked_list(ListNode* head) {
        ListNode* prev_node = NULL;
        ListNode* curr_node = head;
        ListNode* next_node;
        while (curr_node) {
            next_node = curr_node->next;

            curr_node->next = prev_node;
            prev_node = curr_node;
            curr_node = next_node;
        }

        return prev_node;
    }
};
